src/map/RoomGraph.cpp

```diff
@@ -1,77 +1,93 @@
-//
-// Created by simon on 2017-04-30.
-//
-
 #include "RoomGraph.h"
 #include "BossRoom.h"
 #include "HallwayRoom.h"
 
-RoomGraph::RoomGraph(std::function<void(Direction)> walkCallback, std::function<void()> &allPlayersDead) {
-    generateGraph(walkCallback, allPlayersDead);
+RoomGraph::RoomGraph(walk_callback_t walkCallback)
+{
+    generateGraph(walkCallback);
+}
+
+RoomGraph::~RoomGraph()
+{
+    for (int x = 0; x < GRAPH_WIDTH; ++x) {
+        for (int y = 0; y < GRAPH_HEIGHT; ++y) {
+            for (int level = 0; level < GRAPH_LEVELS; ++level) {
+                 graph[x][y][level].reset();
+            }
+        }
+    }
 }
 
-std::shared_ptr<Room> RoomGraph::getCurrentRoom(){
-    return graph[currentX][currentY];
+std::shared_ptr<Room> RoomGraph::getCurrentRoom()
+{
+    return graph[currentX][currentY][currentLevel];
 }
 
-void RoomGraph::walk(Direction direction) {
-    std::pair<int, int> next = getNextFromDir(direction);
-    std::shared_ptr<Room> theRoom = graph[next.first][next.second];
-    if(!theRoom){
+void RoomGraph::walk(Direction direction, win_callback_t winCallback)
+{
+    if (direction == NEXT_LEVEL) {
+        if (currentLevel < GRAPH_LEVELS) {
+            enterNewLevel(currentLevel + 1);
+        } else {
+            winCallback();
+        }
+    }
+
+    level_pos_t next = getNextFromDir(direction);
```

Comment:
Put this in else of (direction == NEXT_LEVEL) and make enterNewLevel return new position instead.